crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -724,13 +769,180 @@ impl RoomSendQueue {
+            if let Some(sizes) = queue.store.lock().await.thumbnail_size_cache.get(related_to) {
+                sizes.get(index).copied().flatten().unwrap_or(0)
+            } else {
+                0
+            }
+        } else {
+            0
+        };
+
+        // If this is a thumbnail upload, get the size of the pending file upload from
+        // the dependent requests.
+        let pending_file_bytes = RoomSendQueue::get_dependent_pending_file_upload_size(
+            queued_request.transaction_id.clone(),
+            queue,
+            room,
+        )
+        .await;
+
+        Some(MediaUploadInfo {
+            index: index as u64,
+            bytes,
+            uploaded_thumbnail_bytes,
+            pending_file_bytes,
+        })
+    }
+
+    /// Determine the size of a pending file upload, if this is a thumbnail
+    /// upload or return 0 otherwise.
+    async fn get_dependent_pending_file_upload_size(
+        txn_id: OwnedTransactionId,
+        queue: &QueueStorage,
+        room: &Room,
+    ) -> usize {
+        let guard = queue.store.lock().await;
+
+        let Ok(client) = guard.client() else {
+            return 0;
+        };
+
+        let Ok(dependent_requests) =
+            client.state_store().load_dependent_queued_requests(room.room_id()).await
+        else {
+            return 0;
+        };
+
+        let Some((cache_key, parent_is_thumbnail_upload)) = dependent_requests.iter().find_map(|r| {
+            if r.parent_transaction_id != txn_id {
+                return None;
+            }
+            as_variant!(&r.kind, DependentQueuedRequestKind::UploadFileOrThumbnail { cache_key, parent_is_thumbnail_upload, .. } => (cache_key.clone(), *parent_is_thumbnail_upload))
+        }) else {
+            return 0;
+        };
+
+        // If this is not a thumbnail upload, we're uploading a gallery and the
+        // dependent request is for the next gallery item.
+        if !parent_is_thumbnail_upload {
+            return 0;
+        }
+
+        if let Ok(cache) = room.client().event_cache_store().lock().await {
+            if let Ok(Some(content)) = cache.get_media_content(&cache_key).await {
+                content.len()
+            } else {
+                0
+            }
+        } else {
+            0
+        }
+    }
+
+    /// Try to create an observable to watch a media's upload progress.
+    fn try_create_media_upload_progress_observable(
+        report_media_upload_progress: &Arc<AtomicBool>,
+        media_upload_info: &MediaUploadInfo,
+        related_txn_id: &Option<OwnedTransactionId>,
+        room_id: &RoomId,
+        global_update_sender: &broadcast::Sender<SendQueueUpdate>,
+        update_sender: &broadcast::Sender<RoomSendQueueUpdate>,
+    ) -> Option<SharedObservable<TransmissionProgress>> {
+        if !report_media_upload_progress.load(Ordering::SeqCst) {
+            return None;
+        }
+
+        if let Some(related_txn_id) = related_txn_id {
```

Comment:
let's avoid one indent level, since this function returns an `Option` :)

`let related_txn_id = related_txn_id?;`